### DISCUSSION TITLE
Completion with trigger character ":" shows no completions

### DISCUSSION THREAD

**Discussion by mdbergmann** [joaotavora/eglot#998] [Q&A] [answered]
Hello.

There is a discussion about this in #841 which I created. It refers to a discussion #740.
In #740 I posted a JSON (it's again below) transcript and a little screencast.

The LSP server used here is lfe-ls (https://github.com/mdbergmann/lfe-ls), it is still rudimentary, however, the issue is persistent.
I use  Emacs 28.1.
Latest Eglot, from git. I update regularly.
No Doom.
     
I'll post the JSON trascript again here:

```
[client-request] (id:20) Tue Feb 15 13:33:17 2022:
(:jsonrpc "2.0" :id 20 :method "textDocument/completion" :params
          (:textDocument
           (:uri "file:///foo.lfe")
           :position
           (:line 149 :character 6)
           :context
           (:triggerKind 2 :triggerCharacter ":")))
[server-reply] (id:20) Tue Feb 15 13:33:17 2022:
(:id 20 :result
     [(:label "columns/0" :kind 3 :detail "io:" :insertText "columns")
      (:label "columns/1" :kind 3 :detail "io:" :insertText "columns")
      (:label "format/1" :kind 3 :detail "io:" :insertText "format")]
```

To reproduce, you can either clone the lfe-ls project, or, if you are on macOS I can provide a build, but still Erlang has to be installed.

Since I use Eglot from Git I tried to see where the problem could be. But after quite a bit of spent time I wasn't able to clearly pinpoint.
Maybe there is a generic issue with using ":" as completion trigger character, because it's used also in Elisp for global symbols (IIRC).

**Comment by joaotavora** [comment]
If there is already an issue for a particular issue, stick to that issue.  If it's a new issue and you don't have the elements to create exactly what's required in the issue template, better create a discussion, as is clearly requested in the issue template.

**Comment by mdbergmann** [comment]
I think it's not necessary to create another discussion. There are now three. Better let me know where the issue is, maybe I can contribute. And then, if there is an issue, it might be worth having a closer look if no one else than the author of Eglot is able to solve it.

**Reply by joaotavora** [reply]
Then why did you create a new issue? Issues, as the template clearly asks, are meant to be created when you have the elements to create them.  If you don't have them don't create an issue , please.  Creating an issue doesn't magically make time in my life for  me to help you with your particular ls or install a haskell toolchain.  I use completion with the `:` all the time.   Try to debug this yourself or make a simple recipe that I or someone else can try.  The simpler the recipe, the greater the chance someone can help you.

**Reply by mdbergmann** [reply]
No problem. I can create a recipe easily. But it requires to install runtimes, LS, etc. Will you follow it? Or is it a waste of my time?
Do you use completion with `:` in Eglot using a trigger character? If yes, I will try to debug.

**Reply by mdbergmann** [reply]
> Do you use completion with : in Eglot using a trigger character? If yes, I will try to debug.

Or, with which language LS do you use this. Maybe I can see if there is any difference in request/response from what my LS does.

**Reply by joaotavora** [reply]
Good idea.  I use it with `clangd` and c++.  `::` is the namespace separator.

But a recipe is even better.  It'd be great is someone :tm: could write an language server simulator that takes a given event log and replays  back the server's responses.  Then we wouldn't have half of this constant "give me a reproduction recipe" dance.

**Reply by mdbergmann** [reply]
Actually Erlang_ls also uses `:` character as triggerCharacter. There completion are showing fine.

So what makes my LS special from LS that also use `:` trigger character is that it is running in a Lisp major mode, LFE is a Lisp, and LFE mode is based on Lisp mode.

I've also tried lsp-mode where completion with `:` character also doesn't work.
However, what I don't understand is that my LS registers two other completion trigger characters: `(` and `'`. Both work fine and show completions. The mechanics in the backend are exactly the same as for `:`.
From this I would conclude that LFE mode / Lisp mode somehow plays a role here.

@joaotavora 
Would you suggest to start debugging `eglot-completion-at-point`? It looks like a complicated function. Would it be worth it?

**Reply by joaotavora** [reply]
> From this I would conclude that LFE mode / Lisp mode somehow plays a role here.

Could be.  Can you link to the source of this "LFE mode"?  Maybe the syntax table designates `:` with some kind of symbol.  What does `(string (char-syntax ?:))` evaluate to in both these modes?

> It looks like a complicated function. Would it be worth it?

That's the thing about debugging, you never know if you're looking in the right spot.  Have you started your emacs bare, naked, without any themes or fancy stuff but Eglot?  That's how I normally progress.  Instead of using company, try doing the thing with the `foo:` and then typing `C-M-i`.  If it fails there then you know the problem isn't the triggering.  If it succeeds, you design another experiment.

**Reply by mdbergmann** [reply]
OK, so I've started bare Emacs, just loaded Eglot, loaded my LFE source code file, but no lfe-mode. I kept it just in fundamental-mode. Attached Eglot to the buffer and voila, all works well, including completion (C-M-i) at something like `(io:`.
Then I did the same in lisp-mode, and there it already doesn't work, no completion options at `(io:`.
If this is an issue with lisp-mode, there is probably little that can be done except working around it.

> What does (string (char-syntax ?:)) evaluate to in both these modes?

I'm not sure with my little Emacs knowhow how to check this.

Update: in scratch buffer this returns "_".

**Comment by joaotavora** [comment]
> Update: in scratch buffer this returns "_".

OK, but this changes (potentially) depending on the major mode. So type `M-: (string (char-syntax ?:))` in both haskell-mode and lfe-mode and tell me what it says.

You still haven't provided a link to lfe-mode...

**Reply by mdbergmann** [reply]
Right, lfe-mode is here: https://github.com/lfe/lfe/blob/develop/emacs/lfe-mode.el

I'll check out the other things.

**Reply by mdbergmann** [reply]
Haskell mode: `.`
Lisp mode and LFE mode: `_`.

Not sure what this means though.

**Reply by joaotavora** [reply]
> Haskell mode: .
> Lisp mode and LFE mode: _.
> Not sure what this means though.

it means that erlang-mode treats `:` as inter-symbol punctuation and lisp mode treat the same character as a symbol constituent.  This means that if you type `M-: (symbol-at-point)` when point is on the `f` of `bar:foo`  you probably get `foo` for erlang-mode and `bar:foo` for lisp-mode.

Which one is correct?  Hard to say, but probably both of them.  However, for the `triggerCharacter` LSP thing to work, the trigger character must be a symbol separator (i.e. punctuation), because `eglot-completion-at-point` completes whole symbols.  That's a limitation of Eglot, but I couldn't come up with a better model of marrying LSP to Emacs completion mechanics.  LSP doesn't have a good way of saying "this is the region where I'm completing".  

Well, there are some hacks, like maybe sweeping that `:detail` field of the completions to try and deduce that.  But not necessarily all servers provide it.

So you might have some luck setting the syntax class of character `:` to `.` (punctuation) in `lfe-mode` 's syntax table.  See the Elisp manual on how to do that.  But that might break other things in `lfe-mode`, such as navigating symbols with `C-M-f` and `C-M-b`.  Maybe that's not very relevant to you, so you might want to try it.

In the meantime I could try other things, but I'm not very confident that they are safe for other servers.

**Reply by mdbergmann** [reply]
Thanks. Setting `:` as punctuation character works.
In fact I think for LFE mode this is how ti should be considering: "or used in a programming language to separate symbols from one another.".

**Comment by joaotavora** [comment, marked as answer]
The reason this doesn't work is that `lfe-mode` has a syntax table that treats `:` as a symbol constituent and `eglot-completion-at-point` completes whole symbols.  To make the `triggerCharacter` LSP thing work in Eglot, the character in question must exist at symbol boundaries.  This is true for `erlang-mode` and `c++-mode` where `:` works well, but not for `lfe-mode`.
